### PR TITLE
feat: changed usage of map for allTsids to Hll

### LINKED ISSUE #1079: [STORY] Use space-efficient sets to track metrics cardinality
# Description
https://github.com/siglens/siglens/pull/1044 added some APIs for metrics cardinality. However, some of those APIs are making large sets to count the number of unique series (TSIDs), and if there's ~1 billion series, that will take a lot of RAM. Instead, we should use a bloom filter or HyperLogLog (HLL) structure to get an estimate of the number of series in a more space-efficient manner. There's examples of bloom filters and HLL elsewhere in the repo.

# (Optional) Testing
Describe how this should be tested and/or which files unit tests should be added to.

For documentation changes or other similar scenarios, this section can be deleted.

# Acceptance Criteria
The APIs still give relatively accurate estimates, but use much less RAM when there's ~1 billion series

### PR DESCRIPTION
**Pull request by dongrep**
# Description
After a bit of research, using Bloom filter to optimize for ~1 Billion series would, take a around 1.5 Gb of space, even if we don't reach that series and for that reason I decided to go with HLL.

I did not test it with a billion series, however I have added a test that succeeds for the map implementation and now for the hll implementation as well.

Fixes #1079 

# Testing
I have added test case that succeeds for the map implementation and now for the hll implementation as well.

# Checklist:
Before marking your pull request as ready for review, complete the following.

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

**Comment by dongrep** [comment]
I would like to collaborate to test exactly how it should be tested.

File: pkg/integrations/prometheus/promql/metricsSearchHandler.go [hunk 1]
```diff
@@ -977,12 +977,14 @@ func ProcessGetMetricSeriesCardinalityRequest(ctx *fasthttp.RequestCtx, myid uin
 				return
 			}
```
[the remark this answers is not in the document]
Comment (dongrep) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go:979]: I have added comments on why error checking has been omitted

**Review by AndrewHess — COMMENTED** [review]
This looks good!

For testing, you can ingest metrics like this [like this](https://github.com/siglens/siglens/blob/develop/.circleci/config.yml#L37). See the [sigcient readme](https://github.com/siglens/siglens/blob/develop/tools/sigclient/README.md) for what the command line args mean; you'll probably need to adjust the values. Then you can use `curl` on from the terminal, or use Postman, or something else, to make a call to `localhost:5122` to [this API](https://github.com/siglens/siglens/blob/develop/pkg/server/query/server.go#L193)

File: pkg/integrations/prometheus/promql/metricsSearchHandler.go [hunk 2]
```diff
@@ -977,12 +977,16 @@ func ProcessGetMetricSeriesCardinalityRequest(ctx *fasthttp.RequestCtx, myid uin
...
 
-			allTsids = utils.MergeMaps(allTsids, tsids)
+			// Err is nil, so tsids is not nil, so we can safely iterate over it
+			for tsid := range tsids {
+				// AddRaw is safe to use because tsid is not nil
+				allTsids.AddRaw(tsid)
+			}
```
[the remark this answers is not in the document]
Comment (dongrep) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go:984]: Thanks for the clarification. I will keep that in mind about Lindy, and make future changes accordingly

[the remark this answers is not in the document]
Comment (dongrep) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go]: I have tested locally by send around 100 million metrics, locally. My machine wasn't able to handle a billion metrics (m1 mac air 8gb) so I tested it with this. 
I tested by comparing both implementations, i.e. map implementation and the hll. Please refer to the screenshots for more info.

Screenshot 2024-09-02 at 5 29 09 PM
Screenshot 2024-09-02 at 5 30 36 PM

 The one with the higher count is the HLL implementation which is expected, but its very close to the map implementation.
 
 I was also watching the activity monitor and it didn't seem to disrupt memory usage, at least not noticeably.

[the remark this answers is not in the document]
Comment (dongrep) [reply pkg/integrations/prometheus/promql/metricsSearchHandler.go]: The time taken for both queries might not be accurate as I was using a debugger for testing it.

**Review by AndrewHess — APPROVED** [review]
Looks good

**Pull request merged**